handler.go

```diff
@@ -78,34 +101,46 @@ var (
 // forms.
 func (srv *Server) Handle(f interface{}) Handler {
 	fv := reflect.ValueOf(f)
-	hf, rt, err := srv.handlerFunc(fv.Type())
+	hf, err := srv.handlerFunc(fv.Type(), nil)
 	if err != nil {
 		panic(errgo.Notef(err, "bad handler function"))
 	}
 	return Handler{
-		Method: rt.method,
-		Path:   rt.path,
+		Method: hf.requestType.method,
+		Path:   hf.requestType.path,
 		Handle: func(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
 			ctx, cancel := contextFromRequest(req)
 			defer cancel()
-			hf(fv, Params{
+			p1 := Params{
 				Response:    w,
 				Request:     req,
 				PathVar:     p,
-				PathPattern: rt.path,
+				PathPattern: hf.requestType.path,
 				Context:     ctx,
-			})
+			}
+			argv, err := hf.unmarshal(p1)
+			if err != nil {
+				srv.WriteError(ctx, w, err)
+				return
+			}
+			hf.call(fv, argv, p1)
 		},
 	}
 }
 
 // Handlers returns a list of handlers that will be handled by the value
-// returned by the given argument, which must be a function of the form:
+// returned by the given argument, which must be a function in one of the
+// following forms:
+//
+// 	func(p httprequest.Params) (T, context.Context, error)
```

Comment:
looking at the standard docs (e.g. https://golang.org/pkg/builtin/#copy) they tend to use "Type" where you have "T". It might be useful to be consistent, although probably not.